src/System.Text.Json/src/System/Text/Json/Reader/JsonReaderHelper.Unescaping.cs

```diff
@@ -0,0 +1,210 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.Buffers.Text;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Text.Json
+{
+    internal static partial class JsonReaderHelper
+    {
+        // Reject any invalid UTF-8 data rather than silently replacing.
+        public static readonly UTF8Encoding s_utf8Encoding = new UTF8Encoding(encoderShouldEmitUTF8Identifier: false, throwOnInvalidBytes: true);
+
+        private const int StackallocThreshold = 256;
+
+        // TODO: Similar to escaping, replace the unescaping logic with publicly shipping APIs from https://github.com/dotnet/corefx/issues/33509
+        public static string GetUnescapedString(ReadOnlySpan<byte> utf8Source, int idx)
+        {
+            byte[] unescapedArray = null;
+            string utf8String;
+
+            Span<byte> unescaped = utf8Source.Length <= StackallocThreshold ?
```

Comment:
Nit: this is working in bytes, which we generally use to mean UTF8, but this is actually UTF16, right?  If so, I'd name the local utf16Unescaped or something like that.

Reply:
Oh, nevermind, it looks like it's still UTF8.  Regardless, I'd still recommend putting that in the name, e.g. utf8Unescaped.